### ISSUE TITLE
[2017 PNAS] Overcoming catastrophic forgetting in neural networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#64]
## 1. Catastrophic Forgetting in Neural Networks

Artificial neural networks due to the tendency for knowledge of previously learnt task(s) (e.g. task A) to be abruptly lost as information relevant to the current task (e.g. task B) is incorporated

This phenomenon, termed **catastrophic forgetting** occurs specifically when the network is trained sequentially on multiple tasks because the weights in the network that are important for task A are changed to meet the objectives of task B

## 2. Intuition

- Recent evidence suggests that the mammalian brain may avoid catastrophic forgetting by protecting previously-acquired knowledge in neocortical circuits
- Neural mechanisms supporting the protection of these strengthened synapses are critical to retention of task performance
- Continual learning in the mammalian neocortex relies on a process of task-specific synaptic consolidation, whereby knowledge about how to perform a previously acquired task is durably encoded in a proportion of synapses that are rendered less plastic and therefore stable over long timescales
- Synaptic consolidation enables continual learning by reducing the plasticity of synapses that are vital to previously learned tasks

### 3.1. Concept
<img width="930" alt="image" src="https://user-images.githubusercontent.com/49643709/166134415-264f8d23-4fcc-475c-b272-9b0c3a7c6942.png">

- Implement an algorithm that performs a similar operation in artificial neural networks by constraining important parameters to stay close to their old values
- Over-parameterization makes it likely that there is a solution for task B, θ_B, that is close to the previously found solution for task A, θ_A
- While learning task B, EWC therefore protects the performance in task A by constraining the parameters to
stay in a region of low error for task A centered around θ_A

### 3.2. Details
<img width="950" alt="image" src="https://user-images.githubusercontent.com/49643709/166134411-a9371b41-cf7c-44e9-95a7-c50945001d65.png">
<img width="945" alt="image" src="https://user-images.githubusercontent.com/49643709/166134428-18ba21ca-a4a2-4706-a64c-4bbd1b6f551f.png">
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/49643709/166140104-521127b3-b08d-408b-863c-64288adea070.png">
<img width="440" alt="image" src="https://user-images.githubusercontent.com/49643709/166140766-7e4fbe48-7a66-4e9c-b0ac-3d15c50a95e3.png">
<img width="754" alt="image" src="https://user-images.githubusercontent.com/49643709/166140134-e9075237-b75a-4a68-a281-35a99fa4c19b.png">

- Posterior를 Laplace approximation을 통해 구함
    - Gaussian distribution with mean given by the parameters
    - Diagonal precision given by the diagonal of the Fisher information matrix F
- Fisher 계산 방법
    1. Dataset을 이용하여 network의 output logit을 구함
    2. Softmax를 통해 확률분포로 바꿔줌
    3. 확률분포와 label을 가지고 negative log-likelihood loss 구함
    4. Loss backward 
    5. 각 parameter별 gradient의 제곱을 더함
    6. 마지막에 데이터 개수로 나눠줌 (len(dataset))
- Numerical 안정성을 위해서 pytorch로 구현할 때 log_softmax + nll로 loss 계산 (cross entropy 써도 됨)
- 즉 기존에 배웠던 task A에 대한 network parameters를 이용하여 미리 저장해서 들고 있음 (self._means, self._precision_matrices)
- 이를 이용하여 중요한 정보를 가지고 있는 (F 값이 큰) parameter는 최대한 덜 변형되게 penalty를 가한다는 것

## 4. Experiments

강화학습과 관련된 실험이 있으나 다루지 않으니 해당 부분은 논문 참고

### 4.1. Dataset

- MNIST dataset
- For each task, generate a fixed, random permutation by which the input pixels of all images would be shuffled
- Each task was thus of equal difficulty to the original MNIST problem, though a different solution would be required for each

### 4.2. Result
<img width="957" alt="image" src="https://user-images.githubusercontent.com/49643709/166134436-543a59a9-e2db-4a31-85f6-fbf7ce802451.png">
<img width="964" alt="image" src="https://user-images.githubusercontent.com/49643709/166134510-4de4a0c4-a49c-451b-8949-8d6457a2cb5a.png">

- Even for the large permutations, the layers of the network closer to the output are indeed being reused for both tasks
- This reflects the fact that the permutations make the input domain very different, but the output domain (i.e. the class labels) is shared
